### PR TITLE
Another diamond trial

### PR DESCRIPTION
**Pull request by Jorge-C**
Wanting travis to build it. Ready for reviews.

### CODE REVIEW

**Comment by Jorge-C** [comment]
For the record, I can reproduce the build error from travis in a docker image with 12.04. Not sure how to further debug it, as the error seems to come from some deep function, and there are about three thousand lines of error output.

**Comment by Jorge-C** [comment]
It's so much faster to debug stuff in a clean docker image rather than pushing and seeing what travis does...

I've identified the issue: diamond does not compile with gcc/g++ 4.6 (version available in travis/ubuntu 12.04). Updating gcc/g++ to 4.7 solves the problem.

**Comment by Jorge-C** [comment]
I got it working on linux/osx! But some osx expert should check that part, as it's using gcc instead of clang on osx.

If it helps someone else figure out future problems, these are the problems/solutions encountered:
- Having a clean (but local) environment to test these things is super important. I'm not sure if there's currently anything better than docker for that, in particular when dealing with system libraries. By being able to reproduce diamond's build error locally, I figured out that diamond needs gcc >= 4.7.
- Updating compilers in travis. If using the old infrastructure with sudo access, it would have been relatively straightforward

```
before_install:
  - sudo add-apt-repository -y ppa:ubuntu-toolchain-r/test
  - sudo apt-get update
  - sudo apt-get install gcc-4.8 g++-4.8
  - sudo update-alternatives --install /usr/bin/gcc gcc /usr/bin/gcc-4.8 60 --slave /usr/bin/g++ g++ /usr/bin/g++-4.8
  - sudo apt-get update -qq
```

However, the newer container-based travis doesn't have sudo access. It's so much faster that it was worth battling with it a bit more. I managed to install gcc 4.8 easily, but for some reason cmake wasn't finding the new compilers, so after many trials I ended up hard coding them. Suggestions welcome!
- The diamond recipe also needs to patch a bug in diamond (reported [in their repo](https://github.com/bbuchfink/diamond/issues/24)).

**Comment by ElDeveloper** [comment]
Just wondering, are you sure you are using gcc and not one of the
symlinks that point to clang?

**Comment by Jorge-C** [comment]
I don't know, that's why I'd like someone who uses osx to double check that part. The following output is coming from cmake in the OSX travis build:

```
-- The C compiler identification is GNU 4.8.3
-- The CXX compiler identification is GNU 4.8.3
```

**Comment by ElDeveloper** [comment]
Got it, thanks for the info. Thinking a bit more about it, it should
just work ... though I could very well be wrong.

File: .travis.yml [hunk 1]
```diff
@@ -8,7 +15,9 @@ env:
...
+    - secure: gKgj4jR2N8WbVBnJCI52zMR21/nd2hWUb05DatiWNJ8pZ+CC6CPD7N/2t8k+y5V25NufI5ZpEBEnGbmV1AxS6ISjPHRejoQgDYTw/PjY4NEIssFVbULIB84FD9SY7/oqkr5L+jrEyDeV52d2LfSAHjwWlL4XYZ7OFdH7HW/5yUj1z4jasfFhv41x8tVknXWAeS1AEdAfwnb4lEeKvlP6jxdfqkQv9PeobM+THV7gGtUo9rA+plZuhQDzXhywRKajEbYbA7/eHzQa5kMYtf8EQl/orDOPCSMEjh9ifCsf0bcFuQkqtOKhgPGVihNdw9PErwhxAckiH2lyELilqbbk10VI/kTKb8y1yP+RmaRxHD1TzdUL5qAxlSQx8IVybnmKCgBW60FKDMsEy4m0AikVch8xkRkhThWFGHjckJFbPIw3m5iwbYK+PEGA43z6lPzUqhHB5Ww8Yeo9xwT26tDyzBhwF5rlHZ63CJ/DxMSQidY+E1ZEmJpN5UqUMxXmM54xAt2fakcUWRH6PY/w984Lc9xY/f667oef4Bgt82N9gjCVVoPERxNnJCgzJShmcnKB6KgTxmqzfI2E2WDqS/MVjoHBgzix04ZU8+gl+BG28HMf3eF3TKHCS4HJlTLZigyZIeuoM241bn+x8zq3WowAh43FtCEb3rmIAG0ur70nzf8=
+    - CC=gcc-4.8
+    - CXX=g++-4.8
```
Comment (RNAer) [inline .travis.yml]: `CC` and `CXX` are still needed? specifying the version globally may interfere the building of other packages?

File: ci/travis/install.sh [hunk 2]
```diff
@@ -21,3 +21,7 @@ conda update -q conda
 conda install pip conda-build anaconda-client
 # Useful for debugging any issues with conda
 conda info -a
+
+# Add channel to get boost libraries. Probably want to "self-host",
+# maybe embedding their recipes as a git module?
+conda config --add channels menpo
```
Comment (RNAer) [inline ci/travis/install.sh:27]: This is not put in `build.sh` because you think `boost` may be required for other packages too?

File: recipes/Diamond/cmake_patch [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+   1.53.0 REQUIRED
+   COMPONENTS program_options timer iostreams thread
+@@ -18,6 +20,6 @@
+   src/basic/options.cpp
+   src/util/tinythread.cpp)
+ 
+-target_link_libraries(diamond blast_core ${Boost_LIBRARIES})
++target_link_libraries(diamond blast_core ${Boost_LIBRARIES} ${CMAKE_THREAD_LIBS_INIT})
+ 
+ install(TARGETS diamond DESTINATION bin)
```
Comment (RNAer) [inline recipes/Diamond/cmake_patch:19]: :+1:

File: ci/travis/install.sh (hunk 2 above)
Comment (Jorge-C) [inline ci/travis/install.sh:27]: Maybe, but it has to be here because there can't be calls to conda inside the build script.

File: .travis.yml (hunk 1 above)
Comment (Jorge-C) [inline .travis.yml]: :+1: I'll drop them

File: ci/travis/install.sh (hunk 2 above)
Comment (RNAer) [inline ci/travis/install.sh:27]: why is that? I think I ran it inside `build.sh`: https://travis-ci.org/biocore/conda-recipes/jobs/90782083#L548

Comment (Jorge-C) [inline ci/travis/install.sh:27]: That build failed. Once I got building working, installing failed because of the conda call. Luckily, it's easy to understand b/c the build script is not called when installing, which means that boost isn't there when trying to run diamond and things fall apart.

tl;dr Nothing can be installed inside build scripts by calling conda, they must be set as dependencies in the meta file.

Comment (Jorge-C) [inline ci/travis/install.sh:27]: Note that this is still suboptimal: to install diamond our users would have to add two channels to their configs, which I think hampers usability. To solve that, we need to either get it working with the default boost, or host the menpo recipes and keep them in the biocore channel (as suggested by Patrick).

Comment (RNAer) [inline ci/travis/install.sh:27]: I see.

Totally agree. If not too much trouble, fixing the default boost is better; otherwise, host menpo recipe as a submodule is also fine.

**Comment by Jorge-C** [comment]
@RNAer I addressed your comments, travis is green.

**Comment by RNAer** [comment]
I thought you'd like to take a stab on the default boost or put menpo boost
as submodule? if not, I can merge it.

**Comment by Jorge-C** [comment]
Yeah, but I won't have time to work on it for a bit so maybe merge and look
into it at a later time. I don't have any strong preference.

**Comment by RNAer** [comment]
There are still problems even it seems built successfully: https://travis-ci.org/biocore/micronota/jobs/91713275

Also error on local Mac:

```
$ conda install -c biocore diamond
$ diamond
dyld: Symbol not found: __ZNK5boost15program_options22error_with_option_name23substitute_placeholdersERKSs
  Referenced from: /Users/zech/anaconda/envs/py3/bin/diamond
  Expected in: /Users/zech/anaconda/envs/py3/bin/../lib//./libboost_program_options.dylib
 in /Users/zech/anaconda/envs/py3/bin/diamond
Trace/BPT trap: 5
```
